src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -293,3 +290,23 @@
       wrap-with-certificate-cn
       wrap-with-default-body
       wrap-with-debug-logging))
+
+(defn wrap-with-parent-check
+  "Middleware that checks the parent exists before serving the rest of the
+   application. This ensures we always return 404's on child paths when the
+   parent data is empty."
+  [app version parent id]
+  (fn [{:keys [globals] :as req}]
+    (let [{:keys [scf-read-db url-prefix]} globals
+          status-fn (case parent
+                      :node nodes/node-exists?
+                      :environment envs/environment-exists?
+                      :report reports/report-exists?)]
+      ;; There is a race condition here, in particular we open up 1 transaction
+      ;; for the parent test, but the rest of the query results are done via the
+      ;; streaming query. This can't be solved until we work out a way to
+      ;; pass through an existing db handle through to the streamed query thread.
```

Comment:
If a user has a node for example, that gets expired/deactivated during the parent exists? query and the data query, it will return an empty list instead of a 404.